core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,218 @@
+    }.toSeq
+
+    val labelArgs = labels.map {
+      case (key, value) => s"$key=$value"
+    } match {
+      case Seq() => Seq()
+      case pairs => Seq("-l") ++ pairs
+    }
+
+    val runArgs = Seq(
+      "run",
+      name,
+      "--image",
+      image,
+      "--generator",
+      "run-pod/v1",
+      "--restart",
+      "Always",
+      "--limits",
+      "memory=256Mi") ++ environmentArgs ++ labelArgs
+
+    runCmd(runArgs, timeouts.run)
+      .map(_ => ContainerId(name))
+  }
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
+    Future {
+      blocking {
+        val pod =
+          kubeRestClient.pods().withName(id.asString).waitUntilReady(timeouts.inspect.length, timeouts.inspect.unit)
+        ContainerAddress(pod.getStatus().getPodIP())
+      }
+    }.recoverWith {
+      case e =>
+        log.error(this, s"Failed to get IP of Pod '${id.asString}' within timeout: ${e.getClass} - ${e.getMessage}")
+        Future.failed(new Exception(s"Failed to get IP of Pod '${id.asString}'"))
+    }
+  }
+
+  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+
+  def logs(id: ContainerId, sinceTime: Option[String])(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    Source.fromFuture(Future {
+      blocking {
+        val path = Path / "api" / "v1" / "namespaces" / kubeRestClient.getNamespace / "pods" / id.asString / "log"
+        val query = Seq("timestamps" -> true.toString) ++ sinceTime.map(time => "sinceTime" -> time)
+        val url = Uri(kubeRestClient.getMasterUrl.toString)
+          .withPath(path)
+          .withQuery(Query(query: _*))
+
+        val request = new Request.Builder().get().url(url.toString).build
+        val response = kubeRestClient.getHttpClient.newCall(request).execute
+        if (response.isSuccessful) {
+          Future.successful(response.body.string)
+        } else {
+          Future.failed(
+            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
+        }
+      }
+    }.flatMap(identity).map { output =>
+      // The k8s logs api has less granularity than the timestamp
+      // we're passing as the 'sinceTime' argument, so we may need to
+      // filter logs from previous activations on a warm runtime pod
+      val original = output.lines.toSeq
+      val relevant = original.dropWhile(s => !s.startsWith(sinceTime.getOrElse("")))
+      val result =
+        if (!relevant.isEmpty && original.size > relevant.size) {
+          // drop matching timestamp from previous activation
+          relevant.drop(1)
+        } else {
+          original
+        }
```

Comment:
Hmm, I'm still not able to grasp this, let me try:

```scala
val original = output.lines
val relevant = sinceTime.map(time => original.dropWhile(!_.startsWith(time)).drop(1)).getOrElse(original)
```

`relevant` should no in any case contain what you need to continue processing, right? The following `if` is redundant, because if relevant is empty, the result should be empty as well?

Reply:
If `relevant` is empty, it means the previous timestamp isn't in `original`. This will be the case for both cold pods and warm pods that haven't been used within the granularity of the timestamp.

Reply:
I should clarify about the warm pods. It's not necessarily that they haven't been used. It's that the log messages from the previous activation are within the server's granularity relative to the activation's last message. This is very rarely more than just the two stream sentinel markers.

Reply:
When I read my comments, I don't feel like they're helping you. :)

Reply:
@markusthoemmes consider most of this thread obsolete; I've done a rewrite with all of this handled as a custom `GraphStage` and it's much saner. It cuts out most of the hackiness of the existing code.